doc/release/upcoming_changes/18880.compatibility.rst

```diff
@@ -0,0 +1,32 @@
+Ufunc ``signature=...`` and ``dtype=`` generalization and ``casting``
+---------------------------------------------------------------------
+The behaviour for ``signature=...`` and ``dtype=...`` can differ in
+some cases to the previous behaviour.
+When ``signature`` was previously used, the casting check on inputs
+was relaxed, which could lead to downcasting inputs unsafely especially
+if combined with ``casting="unsafe"``.
+
+Casting is now guranteed to be safe.  If a signature is only
+partially provided, for example using ``signature=("float64", None, None)``,
+this could lead to no loop being found (an error).
+In that case, it is necessary to provide the complete signature
+to enforce casting the inputs.
+Since NumPy special cases if only outputs (or ``dtype``) is provided,
+this should affect very few users.
```

Comment:
Yeah, that is what I meant, and I think it is probably true :).  It definitely is the mental model I want users to use for now (which will immediately tell them, that practically nobody will notice).

Reply:
:+1:  in that case it might be worth rewording this sentence since IMO it requires a bit of extra context to correctly interpret the "special casing" here. It could just be me though :)